riscv-sbi-prop.md

```diff
@@ -0,0 +1,290 @@
+inter-processor interrupts (IPI) interface, reprogramming timer, serial
+console, and memory barrier instructions. The existing SBI documentation
+can be found here [1]. Many important functionalities such as power
+management/cpu-hotplug are not yet defined due to difficulties in
+accommodating modifications without breaking the backward compatibility
+with the current interface.
+
+The proposed design is inspired by Power State Coordination Interface (PSCI) from
+ARM world. However, it adds only two new mandatory SBI calls providing
+version information and supported Functions, unlike PSCI where a significant
+number of functions are mandatory. The version of the existing SBI will
+be defined as a minimum version(0.1) which will always be backward compatible.
+Similarly, any Linux kernel with a newer feature will fall back if an older
+version of SBI does not support the updated capabilities. Both the operating
+system and SEE can be implemented to be two way backward compatible.
+
+## SBI Calling Conventions (TODO)
+There are some suggestions[3] to distinguish between SBI function return value and
+SBI calling sequence return value. Here are the few possible discussed ideas.
+
+1. Return a single value, but values in the range {-511, -1} are considered
+errors, whereas return values in the range {0, 2^XLEN-512} are considered success.
+
+2. Every SBI function call may return the following structure
+```
+struct valerr {
+  long value;
+  long error;
+};
+```
+According to the RISC-V ABI spec[4], both a0 & a1 registers can be used for function
+return values.
+
+3. Introduce an extra argument that can return error by reference.
+```
+/* Returns the CPU that was started.
+ * @err indicates an error ID if one occurred, otherwise 0.
+ * @err can be NULL if no error detection is required.
+*/
+int start_cpu(int cpu_num,..., Error *err);
+```
+Both approaches (2 & 3) will work, but there will be two versions of existing SBI
+functions to maintain the compatibility.
+
+## SBI Functions:
+
+A SBI function is an individual feature that SBI interface provides to supervisor
+mode from machine mode. Each function ID is assigned as per below section.
+
+#### SBI Function ID numbering scheme:
+A Function Set is a group of SBI functions which collectively implement similar
+kind of feature/functionality. The function ID numbering scheme need to be scalable
+in the future. At the same time, function validation check should be as quick as possible.
+Keeping these two requirements in mind, a hybrid function ID scheme is proposed.
+
+SBI Function ID is u32 type.
+
+```
+31            24                        0
+-----------------------------------------
+|O|            |                        |
+-----------------------------------------
+Function Set   | Function Type          |
+```
+Bit[31]    =  ID overflow bit
+
+Bit[30:24] =  Function Set
+
+Bit[23:0]  =  Function Type within Function Set
```

Comment:
I believe this is a mistake for several reasons, and we can talk if you are interested. You are building on the 1970s era model of a fixed set of well known entry points. This idea was created before we had things like a firmware device tree. You don't need to do what you are doing. You can require that vendors populate the FDT (the FDT they are required to create) with a set of function names and index pairs. The index is used by the handler in the SBI to call the function. The name is used to name the function. At boot, supervisor code can scan the FDT and populate its own table with the name/mapping pair. I've written lots of code like this and I can tell you it is easy.

Reply:
in fact if you do this right, you can actually use the kernel's module loader to link to the sbi, which would be the simplest thing of all. Or you can use the coreboot relocatable module format. 

I'm sorry to be so hard on this idea but this (to me) is a very backwards looking idea on a very forward looking architecture.

Reply:
Having a platform that can be self-probed without passing the right data to it is a benefit. It doesn't mean you can't _also_ come up with a binding that does higher-level descriptions of what is implemented, or a table with handles such that you don't have to probe. But having the API being capable of probing is beneficial, IMHO.

Reply:
It has been remarked that calls into M mode are expensive. This seems to be largely due
to the cost of the context switch. It is desirable, in my view, to be able to do a 
quick SBI call where no state is saved, but a register is loaded with a desired value,
as well as to be able to tell the SBI that it need not save all the state it saves now.

I'd like to propose an alternative calling model for the SBI, aimed at greater efficiency.
This assumes that we can move away from the hard-code SBI numbers model to my proposed
model in which SBI calling information is passed in the FDT, and S mode can create a
mapping of SBI function to call information once booted. This dynamic mapping
allows us to put some firmware-provided attributes into the call to improve efficiency.

To date, SBI call handlers, including the coreboot one, have taken the common approach
that kernels take: trap handler saves state and dispatches to a handler. The handler
then determines what to do based on the trap type. 

The problem is that for very simple register IO, this is a very expensive approach. It's
not currently possible, for example, to do a simple SBI that loads a result into A0,A1 and
returns.

I'd like to structure the SBI call number into two pieces:
-----------------------------
| Call #          | skip #             |
-----------------------------

The Call # will be an SBI call # as before, determined from the FDT.
The skip # will indicate how much
context saving can be skipped. A 0 will indicate none can be skipped;
a 1 will indicate, say,that R1 can be skipped, and so on.

The SBI will provide information about how much saving can be skipped; the caller
can choose to use that information as it wishes. The user might continue to request
that all state be saved; it might request that no state be saved, as it is using
caller save (as in Ken C and Go); or it might use the number provided by the FDT.

The Call # and skip # are not connected. An S-mode binary can call a function,but
save all the needed registers ahead of time, and M mode code might dirty many of
them. Hence
1:31
1:0
call the same SBI call, but S-mode asks for different levels of register saving to occur.

Similarly, the firmware can provide a function and indicate how many registers should
be saved:
MTIMER:0

means that the MTIMER function will dirty no registers and the kernel can, should it wish,
call it with any value between 0 and 31 and get the correct result.

Should the SBI indicate to the kernel that it should save many registers, the kernel
can do this by marking functions as caller-save via a compiler attribute,
or by using a larger skip #. There is more than one way to save context for an SBI call, 
either before the call, or by asking the SBI to do it.

In this way it is possible for the firmware to provide very efficient functions
where no state need be saved; or, for a kernel to arrange a call such that no state
need be saved.

Note that dispatch will now occur in 2 parts. The first, in assembly, will figure out
how much context to save, and save it; the second, in C, will figure out what function
to run.  This will require some rework of how coreboot implements SBI, but now is the
best time.

Note that the skip # of 0, meaning save everything, is chosen for compatibility with
current numbering schemes. Save everything could also be a larger number. 

Finally, I'd suggest the skip # field be 8 bits: more than is needed for now.

Reply:
Per Bruce Hoult's nice note today, this little bit of optimization is probably not what we'll do. We will need to rework but I think his idea is the best.